torch/functional.py

```diff
@@ -324,6 +325,118 @@ def parse_subscript(n: int) -> str:
 
     return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
 
+# If pad specifier is not None and not a tensor, convert it to a tensor
+def _maybe_convert_pad_specifier(pad_spec, arg_name, dtype, device):
+    if pad_spec is None or torch.is_tensor(pad_spec):
+        return pad_spec
+
+    # TODO: torch.tensor(), when called outside of pytorch, throws a warning
+    # if dtype is an integer type and one of the inputs is a float. It would
+    # be nice to use that feature here to detect that case fairly efficiently.
+    # However, for some reason, the warning is not thrown when we call
+    # torch.tensor() here. Something must be disabling the warning, and I'm
+    # not sure how to reenable it.
+    '''
+    import warnings
+
+    with warnings.catch_warnings(record=True) as w_list:
+        pad_spec_tensor = torch.tensor(pad_spec, device=device, dtype=dtype)
+
+    for w in w_list:
+        if isinstance(w.message, DeprecationWarning):
+            if 'an integer is required' in str(w):
+                raise RuntimeError(f"'{arg_name}' must be of integral type")
+
+    return pad_spec_tensor
+    '''
+
+    pad_spec_tensor = torch.tensor(pad_spec, device=device)
+
+    if pad_spec_tensor.dtype != dtype:
+        # If the dtype is integral, make sure the tensor is integral
+        dtype_is_int = (not dtype.is_floating_point) and not (dtype.is_complex)
+
+        if dtype_is_int:
+            pad_spec_is_int = (not pad_spec_tensor.is_floating_point()) and (not pad_spec_tensor.is_complex())
+            assert pad_spec_is_int, f"Expected '{arg_name}' to be of signed integer type"
+
+        pad_spec_tensor = pad_spec_tensor.to(dtype)
+
+    return pad_spec_tensor
+
+
+def pad(input, pad_width, mode='constant', *, constant_values=None, out=None):
+    r"""Pad a tensor by extending the size of one or more dimensions. The new
+    elements are filled with values according to the chosen padding mode.
+    Padding can be added to the beginning of each dimension, called "before
+    padding", and/or the end of each dimension, called "after padding".
+
+    Args:
+        input (Tensor): Tensor to pad
+
+        pad_width (Tensor, sequence, array_like, or int): Number of values
+            padded to the edges of each dimension. If a tensor is given,
+            it must be on CPU. The following formats can be used:
+
+                * Scalar, (all,), or ((all,)) yields equal before and after pad
+                  widths for all dimensions.
+                * ((both_1), ..., (both_N)) yields unique padding for each
+                  dimension, but before and after pad widths within each
+                  dimension are equal.
+                * ((before, after),) or (before, after) yields unique before
+                  and after pad widths, but same for all dimensions.
+                * ((before_1, after_1), ..., (before_N, after_N)) yields unique
```

Comment:
Does there need to be one tuple per dimension of input? 

@saketh-are suggests that nn.functional.pad allows underspecifying padding (not that we want to implement this feature, just that the question may come up)

Reply:
Right, `nn.functional.pad` allows underspecification, but `numpy.pad` does not. If pad_width is a tuple of tuples, `numpy.pad` only accepts it if it contains either 1 tuple (same padding applied to all dimensions) or `ndim` tuples (different padding applied to each dimension). It won't accept a tuple containing greater than 1 and less than `ndim` tuples.

```python
>>> import numpy as np
>>> a = np.ones((2, 2, 2))

# tuple containing 1 tuple works:
>>> np.pad(a, ((1, 1),))
array([[[0., 0., 0., 0.],
...

# tuple containing ndim tuples works:
>>> np.pad(a, ((1, 1), (1, 1), (1, 1)))
array([[[0., 0., 0., 0.],
...

# tuple containing greater than 1 and less than ndim tuples throws error:
>>> np.pad(a, ((1, 1), (1, 1)))
...
ValueError: operands could not be broadcast together with remapped shapes [original->remapped]: (2,2) and requested shape (3,2)
```

We could think about supporting underspecification. But I think that would be confusing from a UX perspective, since a tuple containing 1 tuple would be a weird special case.

Reply:
I agree it's a UX oddity, and if we start by not supporting underspecification we can likely add it later if we change our mind, anyway